### PR TITLE
[CP Staging] Fix opening receipt page from IOU report leads to not found page

### PR DESCRIPTION
**Pull request by bernhardoj**
### Details

The not found page shows because the report ID doesn't match the transaction report ID. This PR fixes it by getting the correct report ID.

### Fixed Issues

$ https://github.com/Expensify/App/issues/39404
PROPOSAL: https://github.com/Expensify/App/issues/39404#issuecomment-2032015584

### Tests

Same as QA Steps
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as QA Steps
### QA Steps

1. Open any chat without unsettled money request
2. Create a new scan receipt request
3. Press the request preview to open the IOU report
4. Press the receipt to open the attachment preview
5. Verify you are able to see the receipt instead of not found page
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/50919443/e163f09d-b750-4874-82d4-3d5bbbff75f1

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/50919443/7c960987-029a-4f78-9280-f89924818d01

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/50919443/68cd1d00-da1c-4056-be15-2edc44639dcb

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/50919443/ba9f736e-5091-4aa5-b3d9-07812bb74ed9

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/50919443/41519ad4-ca7f-42c7-8fa9-9d2b070aa43d

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/50919443/29c5df6b-747f-48e5-aff4-a95226baa12f

</details>

### CODE REVIEW

**Comment by bernhardoj** [comment]
Additional video of:
1. It also works for expense report

https://github.com/Expensify/App/assets/50919443/92b1d6df-1ec4-4c91-9510-d7db0786cb01

2. Track expense still works fine after implementing the fix

https://github.com/Expensify/App/assets/50919443/db38f950-af0f-4a58-a22a-e8cc5698d19b

3. Still works when viewing the receipt from the transaction thread

https://github.com/Expensify/App/assets/50919443/6370456b-ad11-4b56-a47a-df0c8de194a7

**Review by iwiznia — DISMISSED** [review]

**Review by shubham1206agra — CHANGES_REQUESTED** [review]
Please fix this for track expense too

**Comment by bernhardoj** [comment]
Ah yes, I can reproduce it after re-login. The transaction reportID for track expense is set to chat report optimistically, 
https://github.com/Expensify/App/blob/35b4d1b5a91f222986d9a887ecd96ea1ede0f594/src/libs/actions/IOU.ts#L1252-L1255
but looks like BE doesn't save that, so when re-login, the reportID is lost (it becomes '0').

**Comment by bernhardoj** [comment]
Fixed for track expense too.

File: src/pages/TransactionReceiptPage.tsx [hunk 1]
```diff
@@ -45,6 +45,11 @@ function TransactionReceipt({transaction, report, reportMetadata = {isLoadingIni
         // eslint-disable-next-line react-hooks/exhaustive-deps
     }, []);
 
+    const moneyRequestReportID = ReportUtils.isMoneyRequestReport(report) ? report?.reportID : report?.parentReportID;
+    const isTrackExpenseReport = ReportUtils.isTrackExpenseReport(report);
+
+    const shouldShowNotFoundPage = !isTrackExpenseReport && (moneyRequestReportID ?? '') !== transaction?.reportID;
```
[the remark this answers is not in the document]
Comment (bernhardoj) [reply src/pages/TransactionReceiptPage.tsx]: Just updated again, lmk if you have any suggestion

**Comment by bernhardoj** [comment]
fixed lint.

**Review by dukenv0307 — APPROVED** [review]
LGTM, all yours @iwiznia.

File: src/pages/TransactionReceiptPage.tsx [hunk 2]
```diff
@@ -45,6 +45,12 @@ function TransactionReceipt({transaction, report, reportMetadata = {isLoadingIni
         // eslint-disable-next-line react-hooks/exhaustive-deps
     }, []);
 
+    const moneyRequestReportID = ReportUtils.isMoneyRequestReport(report) ? report?.reportID : report?.parentReportID;
+    const isTrackExpenseReport = ReportUtils.isTrackExpenseReport(report);
+
+    // eslint-disable-next-line rulesdir/no-negated-variables
+    const shouldShowNotFoundPage = isTrackExpenseReport ? !transaction : (moneyRequestReportID ?? '') !== transaction?.reportID;
```
Comment (NikkiWines) [inline src/pages/TransactionReceiptPage.tsx:52]: NAB - for brevity 

```suggestion

    // eslint-disable-next-line rulesdir/no-negated-variables
    const shouldShowNotFoundPage = ReportUtils.isTrackExpenseReport(report) ? !transaction : (moneyRequestReportID ?? '') !== transaction?.reportID;
```

**Review by NikkiWines — APPROVED** [review]
:shipit: works great

https://github.com/Expensify/App/assets/16747822/3a513114-3dbf-4f76-9ccc-2c248da5a0ed

**Pull request merged**